### ISSUE TITLE
False positive needless_lifetimes when macro involves

### ISSUE THREAD

**Issue by upsuper** [intellij-rust/intellij-rust#3854]
## Environment

* **Intellij-Rust plugin version:** v0.2.97.2123-191
* **Rust toolchain version:** 1.34.2
* **IDE name and version:** CLion 2019.1.3
* **Operating system:** macOS 10.14.4

## Problem description
The plugin incorrectly suggest needless lifetimes when macro involves, and clippy doesn't.

## Steps to reproduce

```rust
trait Foo {
    type Input;
    fn handle(&self, input: Self::Input);
}

macro_rules! foo_str {
    ($a:lifetime) => { impl Foo<Input = &$a str> }
}

fn handle_foo<'a>(x: foo_str!('a), input: &'a str) {
    x.handle(input);
}
```

In this code, IDE marks the `handle_foo` function `needless_lifetimes`, but Clippy doesn't report so, and indeed the lifetime here is not elidable because it's for an associated type.
